File: telldus-gui/Plugins/Live/livemessagetoken.cpp
#include "livemessagetoken.h"

#include <QDebug>

LiveMessageToken::LiveMessageToken() {
	valueType = Invalid; 
}

LiveMessageToken::LiveMessageToken(const QString &value) {
	valueType = String;
	stringVal = value;
}

QByteArray LiveMessageToken::toByteArray() const {
	if (valueType == Int) {
		return QString("i%1s").arg(intVal, 0, 16).toLocal8Bit();

	} else if (valueType == List) {
		QByteArray retVal("l");
		foreach(LiveMessageToken token, listVal) {
			retVal.append(token.toByteArray());
		}
		retVal.append("s");
		return retVal;

	} else if (valueType == Dictionary) {
		QByteArray retVal("h");
		QHashIterator<QString, LiveMessageToken> it(dictVal);
		while(it.hasNext()) {
			it.next();
			LiveMessageToken key(it.key());
			retVal.append(key.toByteArray());
			retVal.append(it.value().toByteArray());
		}
		retVal.append("s");
		return retVal;
	}
	return QString("%1:%2").arg(stringVal.length(), 0, 16).arg(stringVal).toLocal8Bit();
}

LiveMessageToken LiveMessageToken::parseToken(const QByteArray &string, int* start) {
	LiveMessageToken d;
	if ((*start) >= string.length()) {
		return d;
	}
	if (string[(*start)] == 'i') { //Int
		++(*start);
		int index = string.indexOf('s', (*start));
		if (index < 0) {
			qDebug() << "Error in" << string;
			return d;
		}
		bool ok;
		d.intVal = string.mid((*start), index-(*start)).toInt(&ok, 16);
		if (!ok) {
			return d;
		}
		d.valueType = LiveMessageToken::Int;
		(*start) = index + 1;
		return d;

	} else if (string[(*start)] == 'h') { //Dictionary
		++(*start);
		while((*start) < string.length() && string[(*start)] != 's') {
			LiveMessageToken keyToken(LiveMessageToken::parseToken(string, start));
			if (keyToken.valueType == LiveMessageToken::Invalid) {
				break;
			}
			LiveMessageToken valueToken(LiveMessageToken::parseToken(string, start));
			if (valueToken.valueType == LiveMessageToken::Invalid) {
				break;
			}
			d.valueType = LiveMessageToken::Dictionary;
			d.dictVal[keyToken.stringVal] = valueToken;
		}
		++(*start);
	} else { //String
		int index = string.indexOf(':', (*start));
		if (index < 0) {
			qDebug() << "Error in" << string;
			return d;
		}
		bool ok;
		int length = string.mid((*start), index - (*start)).toInt(&ok, 16);
		if (!ok) {
			return d;
		}
		d.stringVal = string.mid(index+1, length);
		d.valueType = LiveMessageToken::String;
		(*start) = index + length + 1;
	}
	return d;
}

LiveMessageTokenScriptWrapper::LiveMessageTokenScriptWrapper()
	:QObject()
{
}

LiveMessageTokenScriptWrapper::LiveMessageTokenScriptWrapper(LiveMessageToken token)
	:QObject(), p_token(token)
{
}

LiveMessageTokenScriptWrapper::~LiveMessageTokenScriptWrapper() {
}

LiveMessageToken LiveMessageTokenScriptWrapper::token() const {
	return p_token;
}

void LiveMessageTokenScriptWrapper::add(LiveMessageTokenScriptWrapper *t) {
	p_token.valueType = LiveMessageToken::List;
	p_token.listVal << t->p_token;
}

int LiveMessageTokenScriptWrapper::getInt(const QString &key, int defaultValue) {
	if (p_token.valueType != LiveMessageToken::Dictionary) {
		return defaultValue;
	}
	if (!p_token.dictVal.contains(key)) {
		return defaultValue;
	}
	return p_token.dictVal[key].intVal;
}

int LiveMessageTokenScriptWrapper::intVal() const {
	return p_token.intVal;
}

void LiveMessageTokenScriptWrapper::set(const QString &key, int value) {
	p_token.valueType = LiveMessageToken::Dictionary;
	LiveMessageToken token;
	token.valueType = LiveMessageToken::Int;
	token.intVal = value;
	p_token.dictVal[key] = token;
}

void LiveMessageTokenScriptWrapper::set(const QString &key, const QString &value) {
	p_token.valueType = LiveMessageToken::Dictionary;
	p_token.dictVal[key] = LiveMessageToken(value);
}

File: examples/python/live/server/LiveMessageToken.py
# -*- coding: utf-8 -*-

import base64

class LiveMessageToken():
	TYPE_INVALID, TYPE_INT, TYPE_STRING, TYPE_BASE64, TYPE_LIST, TYPE_DICTIONARY = range(6)

	def __init__(self, value = None):
		self.valueType = LiveMessageToken.TYPE_INVALID
		self.stringVal = ''
		self.intVal = 0
		self.dictVal = {}
		self.listVal = []
		if (type(value) is int):
			self.valueType = self.TYPE_INT
			self.intVal = value

		elif (type(value) is str):
			self.valueType = self.TYPE_STRING
			self.stringVal = value

		elif (type(value) is list):
			self.valueType = self.TYPE_LIST
			for v in value:
				self.listVal.append(LiveMessageToken(v))

		elif (type(value) is dict):
			self.valueType = self.TYPE_DICTIONARY
			for key in value:
				self.dictVal[key] = LiveMessageToken(value[key])


	def toByteArray(self):
		if (self.valueType == LiveMessageToken.TYPE_INT):
			return 'i%Xs' % self.intVal

		if (self.valueType == LiveMessageToken.TYPE_LIST):
			retval = 'l'
			for token in self.listVal:
				retval = retval + token.toByteArray()
			return retval + 's'

		if (self.valueType == LiveMessageToken.TYPE_DICTIONARY):
			retval = 'h'
			for key in self.dictVal:
				retval = retval + LiveMessageToken(key).toByteArray() + self.dictVal[key].toByteArray()
			return retval + 's'

		return '%X:%s' % (len(self.stringVal), self.stringVal,)

	@staticmethod
	def parseToken(string, start):
		token = LiveMessageToken()
		if (start >= len(string)):
			return (start, token)

		if (string[start] == 'i'):
			start+=1
			index = string.find('s', start)
			if (index < 0):
				return (start, token)

			try:
				token.intVal = int(string[start:index], 16)
				token.valueType = LiveMessageToken.TYPE_INT
				start = index + 1
			except:
				return (start, token)

		elif (string[start] == 'l'):
			start+=1
			while (start < len(string) and string[start] != 's'):
				start, listToken = LiveMessageToken.parseToken(string, start)
				if (listToken.valueType == LiveMessageToken.TYPE_INVALID):
					break
				token.valueType = LiveMessageToken.TYPE_LIST
				token.listVal.append(listToken)
			start+=1

		elif (string[start] == 'h'):
			start+=1
			while (start < len(string) and string[start] != 's'):
				start, keyToken = LiveMessageToken.parseToken(string, start)
				if (keyToken.valueType == LiveMessageToken.TYPE_INVALID):
					break
				start, valueToken = LiveMessageToken.parseToken(string, start)
				if (valueToken.valueType == LiveMessageToken.TYPE_INVALID):
					break
				token.valueType = LiveMessageToken.TYPE_DICTIONARY
				token.dictVal[keyToken.stringVal] = valueToken
			start+=1

		elif (string[start] == 'u'): #Base64
			start+=1
			start, token = LiveMessageToken.parseToken(string, start)
			token.valueType = LiveMessageToken.TYPE_BASE64
			token.stringVal = base64.decodestring(token.stringVal)

		else: #String
			index = string.find(':', start)
			if (index < 0):
				return (start, token)

			try:
				length = int(string[start:index], 16)
			except:
				return (start, token)

			start = index + length + 1
			token.stringVal = string[index+1:start]
			token.valueType = LiveMessageToken.TYPE_STRING

		return (start, token)

